# This is the code of Tosin OLUYEMI from Nigeria in solution to the trailblazer challenge.

Let's import the necessary modules and data.

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

train_data = pd.read_csv('Train.csv')
test_data = pd.read_csv('Test.csv')
submission_data = pd.read_csv('SampleSubmission.csv')
lasso_1 = pd.read_csv('Lasso_sub_1.csv')

Let's explore the data and make some inferences.

Obviously, we work on the train data first. Let's see the head, columns, missing data, target and understand the data generally.

In [56]:
display(train_data)
display(test_data)
display(submission_data)

Place_ID X Date        Date Place_ID  target  target_min  \
0      010Q650 X 2020-01-02  2020-01-02  010Q650    38.0        23.0   
1      010Q650 X 2020-01-03  2020-01-03  010Q650    39.0        25.0   
2      010Q650 X 2020-01-04  2020-01-04  010Q650    24.0         8.0   
3      010Q650 X 2020-01-05  2020-01-05  010Q650    49.0        10.0   
4      010Q650 X 2020-01-06  2020-01-06  010Q650    21.0         9.0   
...                     ...         ...      ...     ...         ...   
30552  YWSFY6Q X 2020-03-15  2020-03-15  YWSFY6Q    22.0        14.0   
30553  YWSFY6Q X 2020-03-16  2020-03-16  YWSFY6Q    53.0        30.0   
30554  YWSFY6Q X 2020-03-17  2020-03-17  YWSFY6Q    85.0        52.0   
30555  YWSFY6Q X 2020-03-18  2020-03-18  YWSFY6Q   103.0        33.0   
30556  YWSFY6Q X 2020-03-19  2020-03-19  YWSFY6Q    89.0        46.0   

       target_max  target_variance  target_count  \
0            53.0           769.50            92   
1            63.0          1319.85            91   
2            56.0          1181.96            96   
3            55.0          1113.67            96   
4            52.0          1164.82            95   
...           ...              ...           ...   
30552        83.0          3848.86            72   
30553       146.0          9823.87            72   
30554       153.0          8900.85            72   
30555       149.0         13963.90            72   
30556       132.0          9393.64            72   

       precipitable_water_entire_atmosphere  \
0                                 11.000000   
1                                 14.600000   
2                                 16.400000   
3                                  6.911948   
4                                 13.900001   
...                                     ...   
30552                              6.700000   
30553                              6.300000   
30554                              7.100000   
30555                             19.100000   
30556                             11.600000   

       relative_humidity_2m_above_ground  ...  L3_SO2_sensor_zenith_angle  \
0                              60.200001  ...                   38.593017   
1                              48.799999  ...                   59.624912   
2                              33.400002  ...                   49.839714   
3                              21.300001  ...                   29.181258   
4                              44.700001  ...                    0.797294   
...                                  ...  ...                         ...   
30552                          68.300003  ...                   54.302954   
30553                          77.700005  ...                   36.170757   
30554                          68.500000  ...                    8.955216   
30555                          66.300003  ...                   17.974815   
30556                          68.400002  ...                         NaN   

       L3_SO2_solar_azimuth_angle  L3_SO2_solar_zenith_angle  \
0                      -61.752587                  22.363665   
1                      -67.693509                  28.614804   
2                      -78.342701                  34.296977   
3                      -73.896588                  30.545446   
4                      -68.612480                  26.899694   
...                           ...                        ...   
30552                   12.312248                  49.871571   
30553                 -129.236726                  49.779531   
30554                 -162.856069                  48.459205   
30555                 -168.906764                  47.293537   
30556                         NaN                        NaN   

       L3_CH4_CH4_column_volume_mixing_ratio_dry_air  L3_CH4_aerosol_height  \
0                                        1793.793579            3227.855469   
1                                        1789.960449            3384.226562   
2                         

Place_ID X Date        Date Place_ID  \
0      0OS9LVX X 2020-01-02  2020-01-02  0OS9LVX   
1      0OS9LVX X 2020-01-03  2020-01-03  0OS9LVX   
2      0OS9LVX X 2020-01-04  2020-01-04  0OS9LVX   
3      0OS9LVX X 2020-01-05  2020-01-05  0OS9LVX   
4      0OS9LVX X 2020-01-06  2020-01-06  0OS9LVX   
...                     ...         ...      ...   
16131  ZZDJZMS X 2020-03-31  2020-03-31  ZZDJZMS   
16132  ZZDJZMS X 2020-04-01  2020-04-01  ZZDJZMS   
16133  ZZDJZMS X 2020-04-02  2020-04-02  ZZDJZMS   
16134  ZZDJZMS X 2020-04-03  2020-04-03  ZZDJZMS   
16135  ZZDJZMS X 2020-04-04  2020-04-04  ZZDJZMS   

       precipitable_water_entire_atmosphere  \
0                                 11.600000   
1                                 18.300001   
2                                 17.600000   
3                                 15.011948   
4                                  9.700000   
...                                     ...   
16131                              9.400001   
16132                             17.000000   
16133                             19.170641   
16134                              8.400001   
16135                              6.500000   

       relative_humidity_2m_above_ground  specific_humidity_2m_above_ground  \
0                              30.200001                            0.00409   
1                              42.900002                            0.00595   
2                              41.299999                            0.00590   
3                              53.100002                            0.00709   
4                              71.599998                            0.00808   
...                                  ...                                ...   
16131                          86.200005                            0.00413   
16132                          90.300003                            0.00548   
16133                          86.200005                            0.00656   
16134                          66.400002                            0.00457   
16135                          81.599998                            0.00508   

       temperature_2m_above_ground  u_component_of_wind_10m_above_ground  \
0                        14.656824                              3.956377   
1                        15.026544                              4.230430   
2                        15.511041                              5.245728   
3                        14.441858                              5.454001   
4                        11.896295                              3.511787   
...                            ...                                   ...   
16131                     3.136346                              0.073416   
16132                     6.394495                             -1.217649   
16133                     9.750879                             -0.410603   
16134                     8.322565                              1.659441   
16135                     6.857385                             -1.004568   

       v_component_of_wind_10m_above_ground  L3_NO2_NO2_column_number_density  \
0                                  0.712605                          0.000053   
1                                  0.661892                          0.000050   
2                                  1.640559                          0.000050   
3                                 -0.190532                          0.000055   
4                                 -0.279441                          0.000055   
...                                     ...                               ...   
16131                              0.923135                          0.000000   
16132                              1.270308                          0.000000   
16133                              0.754363                          0.000000   
16134                             -2.471616                          0.000000   
16135                              1.752161                          

Place_ID X Date  target
0      0OS9LVX X 2020-01-02       0
1      0OS9LVX X 2020-01-03       0
2      0OS9LVX X 2020-01-04       0
3      0OS9LVX X 2020-01-05       0
4      0OS9LVX X 2020-01-06       0
...                     ...     ...
16131  ZZDJZMS X 2020-03-31       0
16132  ZZDJZMS X 2020-04-01       0
16133  ZZDJZMS X 2020-04-02       0
16134  ZZDJZMS X 2020-04-03       0
16135  ZZDJZMS X 2020-04-04       0

[16136 rows x 2 columns]

Considering the displayed head, 82(77 features and 5 targets) columns in the train data(30557 inputs) and 77 in the test data(16136 inputs).
The target columns are: ['target', 'target_min', 'target_max', 'target_variance', 'target_count'].
Viewing the features, apart from the general readings, there are many considerations of ['NO2', 'O3', 'CO', 'HCHO', 'CLOUD', 'AER_AI', 'SO2', 'CH4']

Most of the columns are float and just 3 columns are non-numerical. 
Be aware that PLACEID X DATE column is a concatenation of both the PLACE ID and DATE columns. It, actually, is not necessary but for submission. Drop it in the training and add it back before submission.
Don't also forget to change the date column from Object to Date.

In [ ]:
display('The shape of the train_data and test data',train_data.shape,'' , test_data.shape)


display('The train data description', train_data.describe())

display('The train columns:', train_data.columns.to_list())

display('The test data columns:', test_data.columns.to_list())

Let's try to clean the data. 
How many missing values and how are they distributed? So many missing data! The maximum data any column lost is 24765 (out of about 30,000, about 81% data lost!)

It seems so much data about CH4 is missing(81% missing in all columns about CH4). That seems to be the highest when compared with 24% of HCHO, 23% of SO2 and 17% of CO. Other columns look negligible

More research has to go into the missing data to know how best to fill them. Something has to be done considering that data is largely missing in the test data too, in the same proportion too.

In [72]:
# Percentage data missing in each column
pct_missing_train = (train_data.isna().sum()/len(train_data.index) * 100).to_dict()
display(pct_missing_train)

pct_missing_test = (test_data.isna().sum()/len(test_data.index) * 100).to_dict()
display(pct_missing_test)


# Exact number of missing data in each column
display(train_data.isna().sum().to_dict())

display(test_data.isna().sum().to_dict())


display(train_data.sort_values(['Date', 'Place_ID']))

{'Place_ID X Date': 0.0,
 'Date': 0.0,
 'Place_ID': 0.0,
 'target': 0.0,
 'target_min': 0.0,
 'target_max': 0.0,
 'target_variance': 0.0,
 'target_count': 0.0,
 'precipitable_water_entire_atmosphere': 0.0,
 'relative_humidity_2m_above_ground': 0.0,
 'specific_humidity_2m_above_ground': 0.0,
 'temperature_2m_above_ground': 0.0,
 'u_component_of_wind_10m_above_ground': 0.0,
 'v_component_of_wind_10m_above_ground': 0.0,
 'L3_NO2_NO2_column_number_density': 7.163661354190529,
 'L3_NO2_NO2_slant_column_number_density': 7.163661354190529,
 'L3_NO2_absorbing_aerosol_index': 7.163661354190529,
 'L3_NO2_cloud_fraction': 7.163661354190529,
 'L3_NO2_sensor_altitude': 7.163661354190529,
 'L3_NO2_sensor_azimuth_angle': 7.166933926759826,
 'L3_NO2_sensor_zenith_angle': 7.166933926759826,
 'L3_NO2_solar_azimuth_angle': 7.166933926759826,
 'L3_NO2_solar_zenith_angle': 7.166933926759826,
 'L3_NO2_stratospheric_NO2_column_number_density': 7.166933926759826,
 'L3_NO2_tropopause_pressure': 7.1669339267598

{'Place_ID X Date': 0.0,
 'Date': 0.0,
 'Place_ID': 0.0,
 'precipitable_water_entire_atmosphere': 0.0,
 'relative_humidity_2m_above_ground': 0.0,
 'specific_humidity_2m_above_ground': 0.0,
 'temperature_2m_above_ground': 0.0,
 'u_component_of_wind_10m_above_ground': 0.0,
 'v_component_of_wind_10m_above_ground': 0.0,
 'L3_NO2_NO2_column_number_density': 8.223847297967279,
 'L3_NO2_NO2_slant_column_number_density': 8.223847297967279,
 'L3_NO2_absorbing_aerosol_index': 8.223847297967279,
 'L3_NO2_cloud_fraction': 8.223847297967279,
 'L3_NO2_sensor_altitude': 8.223847297967279,
 'L3_NO2_sensor_azimuth_angle': 8.223847297967279,
 'L3_NO2_sensor_zenith_angle': 8.223847297967279,
 'L3_NO2_solar_azimuth_angle': 8.223847297967279,
 'L3_NO2_solar_zenith_angle': 8.223847297967279,
 'L3_NO2_stratospheric_NO2_column_number_density': 8.223847297967279,
 'L3_NO2_tropopause_pressure': 8.223847297967279,
 'L3_NO2_tropospheric_NO2_column_number_density': 30.819286068418446,
 'L3_O3_O3_column_number_dens

{'Place_ID X Date': 0,
 'Date': 0,
 'Place_ID': 0,
 'target': 0,
 'target_min': 0,
 'target_max': 0,
 'target_variance': 0,
 'target_count': 0,
 'precipitable_water_entire_atmosphere': 0,
 'relative_humidity_2m_above_ground': 0,
 'specific_humidity_2m_above_ground': 0,
 'temperature_2m_above_ground': 0,
 'u_component_of_wind_10m_above_ground': 0,
 'v_component_of_wind_10m_above_ground': 0,
 'L3_NO2_NO2_column_number_density': 2189,
 'L3_NO2_NO2_slant_column_number_density': 2189,
 'L3_NO2_absorbing_aerosol_index': 2189,
 'L3_NO2_cloud_fraction': 2189,
 'L3_NO2_sensor_altitude': 2189,
 'L3_NO2_sensor_azimuth_angle': 2190,
 'L3_NO2_sensor_zenith_angle': 2190,
 'L3_NO2_solar_azimuth_angle': 2190,
 'L3_NO2_solar_zenith_angle': 2190,
 'L3_NO2_stratospheric_NO2_column_number_density': 2190,
 'L3_NO2_tropopause_pressure': 2190,
 'L3_NO2_tropospheric_NO2_column_number_density': 8669,
 'L3_O3_O3_column_number_density': 283,
 'L3_O3_O3_effective_temperature': 283,
 'L3_O3_cloud_fraction': 283,
 

{'Place_ID X Date': 0,
 'Date': 0,
 'Place_ID': 0,
 'precipitable_water_entire_atmosphere': 0,
 'relative_humidity_2m_above_ground': 0,
 'specific_humidity_2m_above_ground': 0,
 'temperature_2m_above_ground': 0,
 'u_component_of_wind_10m_above_ground': 0,
 'v_component_of_wind_10m_above_ground': 0,
 'L3_NO2_NO2_column_number_density': 1327,
 'L3_NO2_NO2_slant_column_number_density': 1327,
 'L3_NO2_absorbing_aerosol_index': 1327,
 'L3_NO2_cloud_fraction': 1327,
 'L3_NO2_sensor_altitude': 1327,
 'L3_NO2_sensor_azimuth_angle': 1327,
 'L3_NO2_sensor_zenith_angle': 1327,
 'L3_NO2_solar_azimuth_angle': 1327,
 'L3_NO2_solar_zenith_angle': 1327,
 'L3_NO2_stratospheric_NO2_column_number_density': 1327,
 'L3_NO2_tropopause_pressure': 1327,
 'L3_NO2_tropospheric_NO2_column_number_density': 4973,
 'L3_O3_O3_column_number_density': 214,
 'L3_O3_O3_effective_temperature': 214,
 'L3_O3_cloud_fraction': 214,
 'L3_O3_sensor_azimuth_angle': 214,
 'L3_O3_sensor_zenith_angle': 214,
 'L3_O3_solar_azimuth_a

Place_ID X Date        Date Place_ID  target  target_min  \
0      010Q650 X 2020-01-02  2020-01-02  010Q650    38.0        23.0   
94     05EC30X X 2020-01-02  2020-01-02  05EC30X    34.0        17.0   
184    0DPWHX8 X 2020-01-02  2020-01-02  0DPWHX8    73.0        63.0   
278    0GBXTHY X 2020-01-02  2020-01-02  0GBXTHY    17.0         4.0   
372    0HYPV1N X 2020-01-02  2020-01-02  0HYPV1N    87.0         5.0   
...                     ...         ...      ...     ...         ...   
30102  YJENTFL X 2020-04-04  2020-04-04  YJENTFL    26.0        15.0   
30196  YLLOKEY X 2020-04-04  2020-04-04  YLLOKEY    37.0        19.0   
30290  YLZOBFW X 2020-04-04  2020-04-04  YLZOBFW    34.0        12.0   
30384  YPXSK14 X 2020-04-04  2020-04-04  YPXSK14    65.0        38.0   
30478  YSIXKFZ X 2020-04-04  2020-04-04  YSIXKFZ    37.0        23.0   

       target_max  target_variance  target_count  \
0            53.0           769.50            92   
94           95.0          3298.58            23   
184         101.0           846.74            64   
278          59.0          3587.48            21   
372         156.0          7180.03           120   
...           ...              ...           ...   
30102        45.0           880.00            22   
30196        56.0           924.72            72   
30290        78.0          1553.04            89   
30384       112.0          3255.77            72   
30478        61.0           917.07            67   

       precipitable_water_entire_atmosphere  \
0                                 11.000000   
94                                15.837500   
184                               10.800000   
278                               17.600000   
372                               31.800001   
...                                     ...   
30102                              8.300000   
30196                             11.366667   
30290                              4.200000   
30384                              9.820000   
30478                             11.120000   

       relative_humidity_2m_above_ground  ...  L3_SO2_sensor_zenith_angle  \
0                              60.200001  ...                   38.593017   
94                             61.737501  ...                   48.799483   
184                            87.700005  ...                   59.194135   
278                            84.799999  ...                         NaN   
372                            70.300003  ...                    3.698430   
...                                  ...  ...                         ...   
30102                          69.800003  ...                   38.410835   
30196                          92.733332  ...                         NaN   
30290                          76.900002  ...                   46.719428   
30384                          93.760002  ...                         NaN   
30478                          85.540004  ...                         NaN   

       L3_SO2_solar_azimuth_angle  L3_SO2_solar_zenith_angle  \
0                      -61.752587                  22.363665   
94                     -73.607722                  17.462130   
184                   -156.851129                  67.833932   
278                           NaN                        NaN   
372                   -151.127993                  40.672004   
...                           ...                        ...   
30102                 -163.346241                  31.715164   
30196                         NaN                        NaN   
30290                  179.620563                  44.976723   
30384                         NaN                        NaN   
30478                         NaN                        NaN   

       L3_CH4_CH4_column_volume_mixing_ratio_dry_air  L3_CH4_aerosol_height  \
0                                        1793.793579            3227.855469   
94                                               NaN                    NaN   
184                       

Let's separate the target from features;
Drop the first column;
Make the date column a date indeed

No missing data in the target. (Fine!)
Let's shift our focus more to the training features.
Impute the missing values now

In [71]:
train_features = train_data.drop(['Place_ID X Date', 'target', 'target_min', 'target_max', 'target_variance', 'target_count'], axis=1)


train_target = train_data[['target', 'target_min', 'target_max', 'target_variance', 'target_count']]





display(train_features.head())
display(train_target.isna().sum())




Date Place_ID  precipitable_water_entire_atmosphere  \
0  2020-01-02  010Q650                             11.000000   
1  2020-01-03  010Q650                             14.600000   
2  2020-01-04  010Q650                             16.400000   
3  2020-01-05  010Q650                              6.911948   
4  2020-01-06  010Q650                             13.900001   

   relative_humidity_2m_above_ground  specific_humidity_2m_above_ground  \
0                          60.200001                            0.00804   
1                          48.799999                            0.00839   
2                          33.400002                            0.00750   
3                          21.300001                            0.00391   
4                          44.700001                            0.00535   

   temperature_2m_above_ground  u_component_of_wind_10m_above_ground  \
0                    18.516840                              1.996377   
1                    22.546533                              3.330430   
2                    27.031030                              5.065727   
3                    23.971857                              3.004001   
4                    16.816309                              2.621787   

   v_component_of_wind_10m_above_ground  L3_NO2_NO2_column_number_density  \
0                             -1.227395                          0.000074   
1                             -1.188108                          0.000076   
2                              3.500559                          0.000067   
3                              1.099468                          0.000083   
4                              2.670559                          0.000070   

   L3_NO2_NO2_slant_column_number_density  ...  L3_SO2_sensor_zenith_angle  \
0                                0.000156  ...                   38.593017   
1                                0.000197  ...                   59.624912   
2                                0.000170  ...                   49.839714   
3                                0.000175  ...                   29.181258   
4                                0.000142  ...                    0.797294   

   L3_SO2_solar_azimuth_angle  L3_SO2_solar_zenith_angle  \
0                  -61.752587                  22.363665   
1                  -67.693509                  28.614804   
2                  -78.342701                  34.296977   
3                  -73.896588                  30.545446   
4                  -68.612480                  26.899694   

   L3_CH4_CH4_column_volume_mixing_ratio_dry_air  L3_CH4_aerosol_height  \
0                                    1793.793579            3227.855469   
1                                    1789.960449            3384.226562   
2                                            NaN                    NaN   
3                                            NaN                    NaN   
4                                            NaN                    NaN   

   L3_CH4_aerosol_optical_depth  L3_CH4_sensor_azimuth_angle  \
0                      0.010579                    74.481049   
1                      0.015104                    75.630043   
2                           NaN                          NaN   
3                           NaN                          NaN   
4                           NaN                          NaN   

   L3_CH4_sensor_zenith_angle  L3_CH4_solar_azimuth_angle  \
0                   37.501499                  -62.142639   
1                   55.657486                  -53.868134   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   

   L3_CH4_solar_zenith_angle  
0                  22.545118  
1                  19.293652  
2                        NaN  
3                        NaN  
4                        NaN  

[5 rows x 76 columns]

target             0
target_min         0
target_max         0
target_variance    0
target_count       0
dtype: int64

Date                                        0
Place_ID                                    0
precipitable_water_entire_atmosphere        0
relative_humidity_2m_above_ground           0
specific_humidity_2m_above_ground           0
                                        ...  
L3_CH4_aerosol_optical_depth            24765
L3_CH4_sensor_azimuth_angle             24765
L3_CH4_sensor_zenith_angle              24765
L3_CH4_solar_azimuth_angle              24765
L3_CH4_solar_zenith_angle               24765
Length: 76, dtype: int64 Place_ID X Date                             0
Date                                        0
Place_ID                                    0
precipitable_water_entire_atmosphere        0
relative_humidity_2m_above_ground           0
                                        ...  
L3_CH4_aerosol_optical_depth            13021
L3_CH4_sensor_azimuth_angle             13021
L3_CH4_sensor_zenith_angle              13021
L3_CH4_solar_azimuth_angle              13021
L3_CH4_so

Now,Exploring the training features. This is definitely a regression problem. Just saying though.
What does a pivot table look like? Definitely, we will groupby Date vs Place_ID.

Now, we have 94 unique days and 340 unique places.

Won't some visuals sound/look good?
How is the relationship between month and temperature?
How do temperature, precipitable water, relative humidity, specific humidity relate?
Can't we merge the u and v components of windsomehow?
Then, we need to do some magic for each pollutant. Let's merge the train features and the test features before doing feature engineering, though

In [60]:
# drop the place_id x date column in the test data
test_features = test_data.drop(['Place_ID X Date'], axis=1)


# Now, let's merge the feature sets
features = pd.concat([train_features, test_features])
#Turn the date column into datetime object
features['Date'] = pd.to_datetime(features['Date'])



print(features['Date'].dtype)

# Now we can engineer the features as we wish.
# The dates are not so useful, so let's extract the months and drop the exact dates(They all have the same year, so it is irrelevant)
features['Month'] = features['Date'].dt.month
display(features)





datetime64[ns]


Date Place_ID  precipitable_water_entire_atmosphere  \
0     2020-01-02  010Q650                             11.000000   
1     2020-01-03  010Q650                             14.600000   
2     2020-01-04  010Q650                             16.400000   
3     2020-01-05  010Q650                              6.911948   
4     2020-01-06  010Q650                             13.900001   
...          ...      ...                                   ...   
16131 2020-03-31  ZZDJZMS                              9.400001   
16132 2020-04-01  ZZDJZMS                             17.000000   
16133 2020-04-02  ZZDJZMS                             19.170641   
16134 2020-04-03  ZZDJZMS                              8.400001   
16135 2020-04-04  ZZDJZMS                              6.500000   

       relative_humidity_2m_above_ground  specific_humidity_2m_above_ground  \
0                              60.200001                            0.00804   
1                              48.799999                            0.00839   
2                              33.400002                            0.00750   
3                              21.300001                            0.00391   
4                              44.700001                            0.00535   
...                                  ...                                ...   
16131                          86.200005                            0.00413   
16132                          90.300003                            0.00548   
16133                          86.200005                            0.00656   
16134                          66.400002                            0.00457   
16135                          81.599998                            0.00508   

       temperature_2m_above_ground  u_component_of_wind_10m_above_ground  \
0                        18.516840                              1.996377   
1                        22.546533                              3.330430   
2                        27.031030                              5.065727   
3                        23.971857                              3.004001   
4                        16.816309                              2.621787   
...                            ...                                   ...   
16131                     3.136346                              0.073416   
16132                     6.394495                             -1.217649   
16133                     9.750879                             -0.410603   
16134                     8.322565                              1.659441   
16135                     6.857385                             -1.004568   

       v_component_of_wind_10m_above_ground  L3_NO2_NO2_column_number_density  \
0                                 -1.227395                          0.000074   
1                                 -1.188108                          0.000076   
2                                  3.500559                          0.000067   
3                                  1.099468                          0.000083   
4                                  2.670559                          0.000070   
...                                     ...                               ...   
16131                              0.923135                          0.000000   
16132                              1.270308                          0.000000   
16133                              0.754363                          0.000000   
16134                             -2.471616                          0.000000   
16135                              1.752161                          0.000000   

       L3_NO2_NO2_slant_column_number_density  ...  \
0                                    0.000156  ...   
1                                    0.000197  ...   
2                                    0.000170  ...   
3                                    0.000175  ...   
4                                    0.000142  ...   
...                                       ...  .

Okay, let's separate the train data from the test(We have performed a new feature engineering.) And also seize this opportinity to split the train to train, test and validation sets.

In [ ]:
train = features[features.Date.isin(train_features.Date)]
X_train, X_test, y_train, y_test = train_test_split(train, train_target.target, train_size=0.8)
X_train.drop('Date', axis=1)
X_test.drop('Date', axis=1)





# We will need to scale this thing o   
                          
test = features[features.Date.isin(test_features.Date)]
test.drop("Date", axis=1)


Since this is a regression problem, Lasso is a first place to start.

In [ ]:
lasso = Lasso()

abc = cross_val_score(lasso, X_train, y_train, scoring='neg_mean_squared_error', cv=5)
print(abc.best_score_)
print(abc.score(X_test, y_test))

In [ ]:
import sklearn


print(sklearn.metrics.get_scorer_names())